controllers/noderefutil/providerid.go

```diff
@@ -53,24 +53,21 @@ func NewProviderID(id string) (*ProviderID, error) {
 		return nil, ErrEmptyProviderID
 	}
 
-	if !providerIDRegex.MatchString(id) {
-		return nil, ErrInvalidProviderID
-	}
-
-	colonIndex := strings.Index(id, ":")
-	cloudProvider := id[0:colonIndex]
-
-	lastSlashIndex := strings.LastIndex(id, "/")
-	instance := id[lastSlashIndex+1:]
-
 	res := &ProviderID{
 		original:      id,
-		cloudProvider: cloudProvider,
-		id:            instance,
+		cloudProvider: "",
+		id:            "",
 	}
+	if providerIDRegex.MatchString(id) {
+		colonIndex := strings.Index(id, ":")
+		res.cloudProvider = id[0:colonIndex]
+
+		lastSlashIndex := strings.LastIndex(id, "/")
+		res.id = id[lastSlashIndex+1:]
 
-	if !res.Validate() {
-		return nil, ErrInvalidProviderID
+		if !res.Validate() {
+			return nil, ErrInvalidProviderID
+		}
```

Comment:
I have [commented](https://github.com/kubernetes-sigs/cluster-api/issues/8485#issuecomment-1505334369) on the issue as well; if we are going to make this optional, what about just dropping this entirely and make our codebase simpler since what we care about is only equality?

Reply:
That would be a better long-term outcome IMO. Open to others' thoughts as some providers may have opinions on dropping this "validation"?

But: you are 100% correct the only method to fulfill the Cluster API functional contract is equality. Providers should be doing their own validations during node bootstrapping and not relying upon CAPI, but are there any providers who actually rely upon CAPI do do it???

Reply:
@jackfrancis now that we have consensus in https://github.com/kubernetes-sigs/cluster-api/issues/8485, would we great if you can rework the PR based on that. Please suggest if we should open a new PR instead.

Reply:
Hmm, apidiff check noticed that I removed the `Validate()` method. Do we want to change that to something like:

```
// Validate returns true if the provider id is valid.
func (p *ProviderID) Validate() bool {
	return p.String() != ""
}
```

Reply:
Q: The problem for providers is not that they are consuming this code, their problem is that we are using it, right?

What about:
* Deprecating the entire NewProviderID / ProviderID thing
* Introducing a new util if there is still a subset of the functionality that is still useful for external consumers
* Stop using it in our own code altogether (just use a string instead), it seems to provide no benefit as it's just a string equals?

It's a bit of a bigger change, but should not break anyone (?).

I'm probably missing a bunch of things.

Reply:
Providers are actually using it: https://cs.k8s.io/?q=NewProviderID&i=nope&files=&excludeFiles=&repos=

But at least publically there are not that many usages. So we can take a closer look at what they are doing with it.

Reply:
I like @sbueringer idea of CAPI deprecating NewProviderID, nethod, and Core CAPI repository use ProviderId  as a string as it is everywhere and not use the struct. This will not break backward compability as well and and we can remove the NewProviderID once we give some notice.

Reply:
Did slightly more research (https://github.com/search?q=org%3Akubernetes-sigs+NOT+repo%3Akubernetes-sigs%2Fcluster-api+NewProviderID&type=code).

All usages that I could find came ~ down to calling `NewProviderID` and then calling `ID()`. Like this: https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/5195261abef1745a6aa47366c82b378e4161d6ad/cloud/scope/machine.go#L149-L156.

I would suggest for now let's deprecate it and drop our usages (your other PR) and then we can do a follow-up do look at usages across providers and discuss providing a **new** util which covers what providers need (which seems to be parsing ID out of the whole providerID string, although that's "only" useful for providers following a certain format).